sdk/include/opentelemetry/sdk/trace/exporter.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <memory>
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/sdk/trace/recordable.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+/**
+ * ExportResult is returned as result of exporting a span batch.
+ */
+enum class ExportResult
+{
+  /**
+   * Batch was successfully exported.
+   */
+  kSuccess = 0,
+  /**
+   * Exporting failed. The caller must not retry exporting the same batch; the
+   * batch must be dropped.
+   */
+  kFailure
+};
+/**
+ * SpanExporter defines the interface that protocol-specific span exporters must
+ * implement.
+ */
+class SpanExporter
+{
+public:
+  virtual ~SpanExporter() = default;
+
+  /**
+   * Create a span recordable. This object will be used to record span data and
+   * will subsequently be passed to SpanExporter::Export. Vendors can implement
+   * custom recordables or use the default SpanData recordable provided by the
+   * SDK.
+   * @return a newly initialized Recordable object
+   */
+  virtual std::unique_ptr<Recordable> MakeRecordable() noexcept = 0;
```

Comment:
Why is this in the interface? Do I need to implement this function in order to implement an exporter?

Reply:
That's to allow for more efficient implementations.

It's expensive to build up a SpanData-like object that provides an accessor interface. Having this as a customization point, allows for direct serialization approaches where memory is allocated in larger blocks and the serialization is built up eagerly as methods on the tracer are called (I outlined the approach in https://github.com/open-telemetry/opentelemetry-cpp/issues/2)

See, for example, this [benchmark](https://github.com/lightstep/lightstep-tracer-cpp/blob/master/benchmark/manual_tracer_benchmark.cpp) from lightstep's cpp tracer where we compare a custom eager serialization approach (manual) to a SpanData-like appraoch that uses protobuf-generated objects (legacy_manual).  The [results](https://7885-57146219-gh.circle-artifacts.com/0/benchmark/manual_tracer_benchmark_result.txt) show that the cost can be substantially less (ranges from 2x-5x in that run). If you look at the profiles ([manual](https://7885-57146219-gh.circle-artifacts.com/0/benchmark/manual_tracer_benchmark-BM_LoggedSpanReport-manual.svg) vs [legacy_manual](https://7885-57146219-gh.circle-artifacts.com/0/benchmark/manual_tracer_benchmark-BM_LoggedSpanReport-legacy_manual.svg)), you'll see the eager serialization approach saves a lot from memory allocations.

But there's a SpanData Recordable in the SDK, so if an implementer doesn't want to provide their own Recordable, they can write MakeRecordable as a stub that returns a SpanData.